internal/xds/xdsdepmgr/watch_service.go

```diff
@@ -81,3 +90,110 @@ func (r *routeConfigWatcher) stop() {
 		r.depMgr.logger.Infof("Canceling watch on RouteConfiguration resource %q", r.resourceName)
 	}
 }
+
+type clusterWatcher struct {
```

Comment:
Why is it that the listener and route watchers store the cancel func inside themselves, while the cluster and endpoints watchers don't?

Reply:
My preference is to not have the watcher structs do the actual work of starting the watch. So, the dependency manager would create the watcher struct (either literal struct initialization or using a `newXxxWatcher` function) and then start the watch by calling `WatchXxx` on the xDS client. So, the watcher structs will not have the `stop` method. All they would have is three methods corresponding to the resource watcher interface and their implementations would be dead simple, i.e., calling an appropriate method on the dependency manager.

I think this approach has the following benefits:
- the watchers would be dead simple structs
- we *could*, at a later point in time, change them by using generics and reduce the number of watchers from 4 to 1

Let me know if you think otherwise.